src/modules/ekf2/EKF/gps_control.cpp

```diff
@@ -46,187 +46,102 @@ void Ekf::controlGpsFusion(const imuSample &imu_delayed)
-
-		// update GSF yaw estimator velocity (basic sanity check on GNSS velocity data)
-		if ((_gps_sample_delayed.sacc > 0.f) && (_gps_sample_delayed.sacc < _params.req_sacc)
-		    && _gps_sample_delayed.vel.isAllFinite()
-		   ) {
-			_yawEstimator.setVelocity(_gps_sample_delayed.vel.xy(), math::max(_gps_sample_delayed.sacc, _params.gps_vel_noise));
-		}
-	}
+		if  (runGnssChecks(_gps_sample_delayed) && isTimedOut(_last_gps_fail_us, (uint64_t)_min_gps_health_time_us / 2)) {
+			if (isTimedOut(_last_gps_fail_us, (uint64_t)_min_gps_health_time_us)) {
+				// First time checks are passing, latching.
+				_gps_checks_passed = true;
+			}
 
-	if (!gyro_bias_inhibited()) {
-		_yawEstimator.setGyroBias(getGyroBias());
-	}
+			collect_gps(_gps_sample_delayed);
 
-	// run EKF-GSF yaw estimator once per imu_delayed update
-	_yawEstimator.update(imu_delayed, _control_status.flags.in_air && !_control_status.flags.vehicle_at_rest);
+		} else {
+			// Skip this sample
+			_gps_data_ready = false;
 
-	// Check for new GPS data that has fallen behind the fusion time horizon
-	if (_gps_data_ready) {
+			if (_control_status.flags.gps && isTimedOut(_last_gps_pass_us, _params.reset_timeout_max)) {
+				stopGpsFusion();
+				_warning_events.flags.gps_quality_poor = true;
+				ECL_WARN("GPS quality poor - stopping use");
+			}
+		}
 
-		const gpsSample &gps_sample{_gps_sample_delayed};
+	} else if (_control_status.flags.gps) {
+		if (!isNewestSampleRecent(_time_last_gps_buffer_push, _params.reset_timeout_max)) {
+			stopGpsFusion();
+			_warning_events.flags.gps_data_stopped = true;
+			ECL_WARN("GPS data stopped");
+		}
+	}
 
-		const bool gps_checks_passing = isTimedOut(_last_gps_fail_us, (uint64_t)5e6);
-		const bool gps_checks_failing = isTimedOut(_last_gps_pass_us, (uint64_t)5e6);
+	if (_gps_data_ready) {
+		const gnssSample &gnss_sample = _gps_sample_delayed;
 
 #if defined(CONFIG_EKF2_GNSS_YAW)
-		controlGpsYawFusion(gps_sample, gps_checks_passing, gps_checks_failing);
+		controlGpsYawFusion(gnss_sample);
 #endif // CONFIG_EKF2_GNSS_YAW
 
-		// GNSS velocity
-		const Vector3f velocity{gps_sample.vel};
-		const float vel_var = sq(math::max(gps_sample.sacc, _params.gps_vel_noise));
-		const Vector3f vel_obs_var(vel_var, vel_var, vel_var * sq(1.5f));
-		updateVelocityAidSrcStatus(gps_sample.time_us,
-					   velocity,                                                   // observation
-					   vel_obs_var,                                                // observation variance
-					   math::max(_params.gps_vel_innov_gate, 1.f),                 // innovation gate
-					   _aid_src_gnss_vel);
-		const bool gnss_vel_enabled = (_params.gnss_ctrl & static_cast<int32_t>(GnssCtrl::VEL));
-
-		// GNSS position
-		const Vector2f position{gps_sample.pos};
-		// relax the upper observation noise limit which prevents bad GPS perturbing the position estimate
-		float pos_noise = math::max(gps_sample.hacc, _params.gps_pos_noise);
+		updateGnssVel(gnss_sample, _aid_src_gnss_vel);
```

Comment:
Simply move this earlier (like line ~62) so that estimator_aid_src_gnss_vel is always updated?

Reply:
Good idea, I moved them.